extension/src/state_aggregate.rs

```diff
@@ -0,0 +1,618 @@
+    flatten,
+    palloc::{Inner, Internal},
+    pg_type,
+    raw::{bytea, TimestampTz},
+    ron_inout_funcs,
+};
+
+use toolkit_experimental::StateAgg;
+
+#[pg_schema]
+pub mod toolkit_experimental {
+    use super::*;
+
+    pg_type! {
+        #[derive(Debug)]
+        struct StateAgg<'input> {
+            states_len: u64,
+            durations_len: u64,
+            durations: [DurationInState; self.durations_len],
+            states: [u8; self.states_len],
+        }
+    }
+
+    impl StateAgg<'_> {
+        pub fn new(states: String, durations: Vec<DurationInState>) -> Self {
+            let states_len = states.len() as u64;
+            let durations_len = durations.len() as u64;
+            unsafe {
+                flatten!(StateAgg {
+                    states_len,
+                    states: states.into_bytes().into(),
+                    durations_len,
+                    durations: (&*durations).into(),
+                })
+            }
+        }
+
+        pub fn get(&self, state: &str) -> Option<i64> {
+            for record in self.durations.iter() {
+                if self.state_str(&record) == state {
+                    return Some(record.duration);
+                }
+            }
+            None
+        }
+
+        fn state_str(&self, record: &DurationInState) -> &str {
+            let beg = record.state_beg as usize;
+            let end = record.state_end as usize;
+            let state = &self.states.as_slice()[beg..end];
+            // SAFETY: came from a String in `new` a few lines up
+            unsafe { std::str::from_utf8_unchecked(state) }
+        }
+    }
+
+    ron_inout_funcs!(StateAgg);
+}
+
+#[aggregate]
+impl toolkit_experimental::state_agg {
+    type State = StateAggTransState;
+
+    const PARALLEL_SAFE: bool = true;
+
+    fn transition(
+        state: Option<State>,
+        #[sql_type("timestamptz")] ts: TimestampTz,
+        #[sql_type("text")] value: Option<String>,
+    ) -> Option<State> {
+        let value = match value {
+            None => return state,
+            Some(value) => value,
+        };
+        let mut state = state.unwrap_or_else(StateAggTransState::new);
+        state.record(value, ts.into());
+        Some(state)
+    }
+
+    fn combine(a: Option<&State>, b: Option<&State>) -> Option<State> {
+        #[cfg(any(test, feature = "pg_test"))]
+        {
+            use std::sync::atomic::Ordering::Relaxed;
+            use tests::counters::*;
+            match (a, b) {
+                (None, None) => COMBINE_NONE.fetch_add(1, Relaxed),
+                (Some(_), None) => COMBINE_A.fetch_add(1, Relaxed),
+                (None, Some(_)) => COMBINE_B.fetch_add(1, Relaxed),
+                (Some(_), Some(_)) => COMBINE_BOTH.fetch_add(1, Relaxed),
+            };
+        }
+        match (a, b) {
+            (None, None) => None,                     // TODO untested
+            (Some(only), None) => Some(only.clone()), // TODO untested
```

Comment:
It would be nice to be able to reach these arms, but I don't think it needs to block merge.